### PR TITLE
Process WASM blocks from substreams

### PR DESCRIPTION
**Pull request by mangas**
Add support for running WASM mappings on incoming substreams blocks. 
- The block stream will emit the encoded result as a byte array
- The mapping will need to handle the decoding of the input

Example is provided [here](https://github.com/mangas/near-wasm-block)

### CODE REVIEW

File: core/src/subgraph/runner.rs [hunk 1]
```diff
@@ -880,197 +1077,247 @@ where
     C: Blockchain,
     T: RuntimeHostBuilder<C>,
 {
-    async fn handle_process_block(
+    async fn handle_process_wasm_block(
```
Comment (lutter) [inline core/src/subgraph/runner.rs:1080]: There seems to be _a lot_ of code duplication between `handle_process_wasm_block` and `handle_process_block` which I fear will lead to subtle bugs as that logic evolves over time. Is there no way to avoid that duplication? I think `handle_process_block` would also be improved by introducing some helper structs that encapsulate some of the logic that is shared between the two; I know it's grown over time, but to the uninitiated, the code for `handle_process_block` is very hard to follow.

**Review by lutter — CHANGES_REQUESTED** [review]
The biggest thing that I'd like to see addressed is the code duplication between `handle_process_block` and `handle_process_wasm_block`.

Since this touches and refactors some very sensitive core logic, it also needs a run in the integration cluster.

File: core/src/subgraph/context.rs [hunk 2]
```diff
@@ -114,6 +115,62 @@ impl<C: Blockchain, T: RuntimeHostBuilder<C>> IndexingContext<C, T> {
...
+            proof_of_indexing
+                .borrow_mut()
+                .start_handler(causality_region);
+        }
+
+        let start = Instant::now();
+        state = self
+            .instance
+            .hosts()
+            .first()
```
[the remark this answers is not in the document]
Comment (mangas) [reply core/src/subgraph/context.rs:143]: Hosts exist map 1:1 to DataSources, on substreams we only allow one data source since it's triggered on every block so we basically only care about the first one. 

The code around the runner has all these points where context is required to understand what is going. It's something I would like to refactor at some point in the near future.

File: chain/arweave/src/chain.rs [hunk 3]
```diff
@@ -280,12 +277,13 @@ impl SubstreamsMapper<Chain> for FirehoseMapper {
...
         &self,
         _logger: &Logger,
-        _clock: &Clock,
-        _block: &prost_types::Any,
-    ) -> Result<BlockWithTriggers<Chain>, Error> {
+        _clock: Clock,
+        _cursor: FirehoseCursor,
+        _block: Vec<u8>,
+    ) -> Result<BlockStreamEvent<Chain>, Error> {
         unimplemented!()
```
[the remark this answers is not in the document]
Comment (mangas) [reply chain/arweave/src/chain.rs:287]: If this code is reached it is a problem. Substreams is itself a chain so anything that calls into this code should be on another chain

File: core/src/subgraph/context.rs (hunk 2 above)
[the remark this answers is not in the document]
Comment (mangas) [reply core/src/subgraph/context.rs:143]: Added a comment

File: core/src/subgraph/runner.rs (hunk 1 above)
Comment (lutter) [reply core/src/subgraph/runner.rs:1080]: Can you then add an issue so this doesn't get completely forgotten?

**Review by lutter — APPROVED** [review]

**Comment by lutter** [comment]
I've approved this though I have pretty big reservations about the code duplication, and the fact that that code is pretty much undocumented (what's the high-level intended behavior of that code?) and very hard to follow.

File: core/src/subgraph/runner.rs [hunk 4]
```diff
@@ -698,6 +727,157 @@ where
...
+                }
+
+                if let Some(stop_block) = &self.inputs.stop_block {
+                    if block_ptr.number >= *stop_block {
+                        info!(self.logger, "stop block reached for subgraph");
+                        return Ok(Action::Stop);
+                    }
+                }
+
+                if matches!(action, Action::Restart) {
```
[the remark this answers is not in the document]
Comment (mangas) [reply core/src/subgraph/runner.rs:793]: oh I think I see the issue, this is all old code, moved from elsewhere in the runner so it can be re-used so Idont want to change anything here
